libdrgn/iterators.c

```diff
@@ -0,0 +1,461 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "iterators.h"
+#include "helpers.h"
+#include "type.h"
+
+#define RADIX_TREE_ENTRY_MASK ((uint64_t)3)
+
+DEFINE_VECTOR_FUNCTIONS(frames)
+
+static struct drgn_error *
+radix_tree_iter_run_frame(struct radix_tree_iter *iter, bool *is_exhausted);
+
+static struct drgn_error *
+radix_tree_iter_handle_node(struct radix_tree_iter *iter,
+			    struct drgn_object *node, uint64_t index)
+{
+	struct drgn_error *err;
+	struct radix_tree_iter_frame *frame = NULL;
+	uint64_t value;
+
+	err = drgn_object_read(node, node);
+	if (err)
+		goto out;
+	err = drgn_object_read_unsigned(node, &value);
+	if (err)
+		goto out;
+	if (((value & RADIX_TREE_ENTRY_MASK) !=
+	     iter->RADIX_TREE_INTERNAL_NODE)) {
+		// Base-case, node is NOT internal
+		iter->entry.node = *node;
+		iter->entry.index = index;
+		return NULL;
+	}
+
+	// We are dealing with an internal node, and must iterate over its slots
+
+	err = drgn_object_set_unsigned(node, iter->node_type,
+				       value & ~iter->RADIX_TREE_INTERNAL_NODE,
+				       0);
+	if (err)
+		goto out;
+	frame = frames_append_entry(&iter->frames);
+	if (!frame) {
+		err = &drgn_enomem;
+		goto out;
+	}
+	frame->index = index;
+	frame->loop_index = 0;
+	struct drgn_object tmp;
+	drgn_object_init(&tmp, iter->prog);
+	drgn_object_init(&frame->slots, iter->prog);
+	err = drgn_object_member_dereference(&tmp, node, "shift");
+	if (err)
+		goto out;
+	err = drgn_object_read_unsigned(&tmp, &frame->shift);
+	if (err)
+		goto out;
+	err = drgn_object_member_dereference(&frame->slots, node, "slots");
+	if (err)
+		goto out;
+
+out:
+	if (err) {
+		if (frame) {
+			drgn_object_deinit(&frame->slots);
+			frames_pop(&iter->frames);
+		}
+		return err;
+	}
+
+	bool _;
+	return radix_tree_iter_run_frame(iter, &_);
```

Comment:
I was able to completely get rid of the recursion with only relatively minor tweaks to my existing code (instead of a full-scale re-write following the above example), let me know if the result is satisfactory.